File: TP01/RecPalindromo.java
class RecPalindromo{
    public static boolean isFim(String s){   // Testa se a palavra digitada foi "FIM"
        return (s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
    }

    public static boolean isPalindromo (String s){
        return isPalindromo(s, 0);}

    public static boolean isPalindromo (String palavra, int i){  // Retorna true ou false para palindromo
        boolean isP = true;
        int tam = palavra.length()-1;

        if(i < tam/2){
            if(palavra.length() < 2){
            isP = true;
        } 
            else{
            if(palavra.charAt(i) == palavra.charAt(tam-i)){
            isP = true;
            isP = isPalindromo(palavra, i+1);
        } 
            else{
            isP = false;
        }
        }
    }
    return isP;
    }

    public static void main (String[] args){
        String[] entrada = new String[1000];
        int numEntrada = 0;

        //Leitura da entrada padrao 
        do {
            entrada[numEntrada] = MyIO.readLine();
        } while (isFim(entrada[numEntrada++]) == false);
        numEntrada--;   //Desconsiderar ultima linha contendo a palavra FIM

       //Para cada linha de entrada, gerando uma de saida boleana se sim ou nao palindromo
        for(int i = 0; i < numEntrada; i++){
            if(isPalindromo(entrada[i])){
                MyIO.println("SIM");
            } else {
                MyIO.println("NAO");
            }
        }
    }
}

File: TP01/RecPalindromo.c
#include<stdio.h>
#include<stdbool.h>
#include<string.h>

bool isPalindromo (char palavra[], int i, int a){  // Retorna true ou false para palindromo
    bool isP = true;

    if(a < (i/2)){
        if(palavra[a] != palavra[i-1-a]){
            isP = false;
            } 
        else{
            isP = true;
            isP =isPalindromo(palavra, i, a+1);
            }
    }
    return(isP);   
}

bool isFim(char s[], int i){  // Testa se a palavra digitada foi "FIM"
    return (i == 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

int main( ){
    int tam = 0;
    char palavra[1000];
    int i = 0;

    do{
        fgets(palavra, 1000, stdin);
        tam = strlen(palavra)-1; 
        
        if (!isFim(palavra, tam))
        {
            if (isPalindromo(palavra, tam, 0))
            {
                printf("SIM\n");
            } 
            else
            {
                printf("NAO\n");
            } 
        }
    }while (!isFim(palavra, tam));
return 0;
}